core/unit_test/view/TestReferenceCountedDataHandle.hpp

```diff
@@ -0,0 +1,131 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+
+#include <gtest/gtest.h>
+
+/*
+template <class ElementType, class MemorySpace>
+class ReferenceCountedDataHandle {
+ public:
+  using value_type   = ElementType;
+  using pointer      = value_type*;
+  using reference    = value_type&;
+  using memory_space = MemorySpace;
+  ...
+}
+*/
+
+namespace {
+using element_t = float;
+using mem_t     = typename TEST_EXECSPACE::memory_space;
+using data_handle_t =
+    Kokkos::Impl::ReferenceCountedDataHandle<element_t, mem_t>;
+using const_data_handle_t =
+    Kokkos::Impl::ReferenceCountedDataHandle<const element_t, mem_t>;
+;
+}  // namespace
+
+TEST(TEST_CATEGORY, RefCountedDataHandle_Typedefs) {
+  Kokkos::parallel_for(
+      Kokkos::RangePolicy<TEST_EXECSPACE>(0, 1), KOKKOS_LAMBDA(int) {
```

Comment:
Do you really need to write those checks within a parallel region ? Sounds awkward.

Reply:
Yes and no. In principle it is possible for typedefs to differ inside Device code and Host code. I mean it would be extremely stupid to do that but it is possible. This was we test for it.

Reply:
I should add that its likely an ODR violation if one were to do that ...

Reply:
https://godbolt.org/z/fn7erb77v

its somewhat contrived obviously, but on the other hand the way we end up with the typedefs is complicated enough that I don't think its impossible to end up with weird stuff.

Reply:
OK, never mind.

Reply:
> In principle it is possible for typedefs to differ inside Device code and Host code

The code snippet you posted is highly unrealistic/artificial.  We are testing a Kokkos class template that may be templated on user-defined types but I fail to see for what naughty definition this test would fail (unless they specialize `std::is_same_v` in which case a/ I don't think we should care b/ your test would also break)